File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    
    const idxMap = new Map(players.map((p, i) => [p, i])); 
    
    for(const calling of callings){
        
        
        // 인덱스 조회 O(1)
        const idx = idxMap.get(calling);
        
        const before = players[idx - 1];
        
        players[idx - 1] = calling;
        players[idx] = before;
        
        idxMap.set(calling, idx-1);
        idxMap.set(before, idx);
        
    }
    
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    result = players
    
    players_dict = {p: i for i, p in enumerate(players)}
    
    for calling in callings:
        cur_rank = players_dict[calling]
        prev_rank = cur_rank - 1
        prev_player = result[prev_rank]
        result[prev_rank] = calling
        result[cur_rank] = prev_player
        
        players_dict[calling] = prev_rank
        players_dict[prev_player] = cur_rank
    
    return result